### DISCUSSION TITLE
l10n and webpack

### DISCUSSION THREAD

**Discussion by worksofliam** [microsoft/vscode-discussions#517] [Extension Development QnA] [answered]
Hi folks

Based on this repo: https://github.com/microsoft/vscode-extension-samples/tree/main/l10n-sample

1. I compile my extension with webpack. How does this affect the `l10n` files? I am guessing webpack won't pack them up and then they will be shipped as many separate files in the extension?

2. Do I have to have the `package.nls` files in the root of the extension repo? Can they also belong in a subdirectory?

3. Something the example doesn't message is what language it will fallback to. Here it mentions that a translation for JP is available, but doesn't mention about what the fallback is.

> Localized strings will be pulled out of a bundle.l10n.<LANG>.json file if one is available. In this repo, [we have one for Japanese](https://github.com/microsoft/vscode-extension-samples/blob/main/l10n-sample/l10n/bundle.l10n.ja.json).

edit: I think I figured number 3 out. Basically you pass in the entire base string into `.t` and it will default to that if no translation is available I guess? Not too good if the strings are huge?

Thanks!

**Comment by alefragnani** [comment, marked as answer]
Hi,

I have embraced `l10n` on all of my extensions, so I'll try to help based on what I have experienced

> 1. I compile my extension with webpack. How does this affect the l10n files? I am guessing webpack won't pack them up and then they will be shipped as many separate files in the extension?

No, it won't affect. The `l10n` folder will be available in extension's root folder, just like the `package.nls.xxx.json` files are there today.

```
   CHANGELOG.md
   dist
      bundle.l10n.json
      bundle.l10n.pt-br.json
   LICENSE.md
   ...
   package.json
   package.nls.json
   package.nls.pt-br.json
   ...
```

> 2. Do I have to have the package.nls files in the root of the extension repo? Can they also belong in a subdirectory?

These are two different _technologies_, and you can have `l10n` but not `package.nls.xxx.json`, and vice-versa. AFAIK, `package.nls.json` files cannot be inside a folder. At least, I didn't find any documentation about it.

> 3. Something the example doesn't message is what language it will fallback to. Here it mentions that a translation for JP is available, but doesn't mention about what the fallback is.

You figured out correctly. You pass the fallback/default string to `.t`. BTW, you don't even need a `bundle.l10n.json` file inside `l10n` folder, just the _other languages_

Hope this helps

**Reply by worksofliam** [reply]
@alefragnani thanks for the great reply.

Regarding question 2, perhaps we should make an issue so the `package.nls.json` files (translations for `package.json`) aren't required in the root. It seems like a lot of clutter to me.

Also, after having a discussion with a team member, I'm not sure how much we like the l10n API, so we may or may not roll our own.

**Reply by alefragnani** [reply]
Thank you, I'm glad to help!

Well, I really liked the new `l10n` API, and its tooling. IMHO, it is way better than the original `i18n` API, specially if you work with non OSS extensions. 

If you miss something, take a look at its repo (https://github.com/microsoft/vscode-l10n) for opened issues. Maybe it is already there. I have reported a couple of issues and asked another couple of feature requests, and it is being a great experience. In fact, even already embracing the API on all of my extensions, I'm planning a few improvements on it.

**Reply by worksofliam** [reply]
My primary concern with `t` is that the first parameter is the entire base message (which also acts as the if), where I would prefer it to just be an id.

I personally was hoping there would be a base translation file, and if the translation for the current language being used doesn't exist, it would fall back to the base.

Instead, the base is the first parameter of `t` which is also the id/key for every translation file.

**Reply by alefragnani** [reply]
I see. 

Well, for regular localizations (with no parameters), nothing blocks you from using it. Simply use an `id` instead and make your _base_ `./l10n/bundle.l10n.json` file to be:

```json
{ 
  "some-id":"Fallback Value"
} 
```
instead of  
```json
{ 
  "Fallback Value": "Fallback Value"
}
```

On the other hand, if you need to use parameters on your strings, it won't be as _kind of id_ as you expect.

**Reply by worksofliam** [reply]
Ah ok, so there is a way to have a base file, that's good.

But I see your point that if I use 'id's then I won't be able to use parameters.

Thanks for clearing that up for me!

**Reply by alefragnani** [reply]
Yep, there are base files. 

When you run the `@vscode/l10n-dev` CLI on your source, it will extract all `.t` calls you have to create the base `bundle.l10n.json` file. Being the fallback, you will notice that both key and value are the same, for every entry.
